cmd/fs-v1-metadata.go

```diff
@@ -274,167 +260,83 @@ func newFSMetaV1() (fsMeta fsMetaV1) {
-		entriesSet := set.CreateStringSet(entries...)
-		expectedConfigsSet := set.CreateStringSet(expectedConfigs...)
-
-		// Entries found shouldn't be more than total
-		// expected config directories, files.
-		if len(entriesSet) > len(expectedConfigsSet) {
-			return errCorruptedFormat
-		}
-
-		// Look for the difference between entries and the
-		// expected config set, resulting entries if they
-		// intersect with original entries set we know
-		// that the backend has unexpected files.
-		if !entriesSet.Difference(expectedConfigsSet).IsEmpty() {
-			return errCorruptedFormat
-		}
+	// Check format FS.
+	if err = format.CheckFS(); err != nil {
+		rlk.Close()
+		return nil, err
 	}
-	return nil
-}
 
-// Check for sanity of FS format for a given version.
-func checkFormatSanityFS(fsPath string, fsFormatVersion string) (err error) {
-	switch fsFormatVersion {
-	case fsFormatV2:
-		err = checkFormatSanityFSV2(fsPath)
-	default:
-		err = errCorruptedFormat
-	}
-	return err
+	//  Always return read lock here and should be closed by the caller.
+	return rlk, traceError(err)
 }
 
-// Initializes a new `format.json` if not present, validates `format.json`
-// if already present and migrates to newer version if necessary. Returns
-// the final format version.
 func initFormatFS(fsPath, fsUUID string) (err error) {
-	fsFormatPath := pathJoin(fsPath, minioMetaBucket, fsFormatJSONFile)
-
-	// fsFormatJSONFile - format.json file stored in minioMetaBucket(.minio.sys) directory.
-	lk, err := lock.LockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
-	if err != nil {
-		return traceError(err)
-	}
-	defer lk.Close()
-
-	var format = &formatConfigV1{}
-	_, err = format.ReadFrom(lk)
-	// For all unexpected errors, we return.
-	if err != nil && errorCause(err) != io.EOF {
-		return traceError(fmt.Errorf("Unable to load 'format.json', %s", err))
-	}
-
-	// If we couldn't read anything, The disk is unformatted.
-	if errorCause(err) == io.EOF {
-		err = errUnformattedDisk
-		format = newFSFormatV2()
-	} else {
-		// Validate loaded `format.json`.
-		err = checkFormatFS(format, fsFormatVersion)
-		if err != nil && err != errFSFormatOld {
-			return traceError(fmt.Errorf("Unable to validate 'format.json', %s", err))
-		}
-	}
-
-	// Disk is in old format migrate object metadata.
-	if err == errFSFormatOld {
-		if merr := migrateFSObject(fsPath, fsUUID); merr != nil {
-			return merr
+	fsFormatPath := pathJoin(fsPath, minioMetaBucket, formatConfigFile)
+
+	// This loop validates format.json by holding a read lock and
+	// proceeds to holding a TryWriteLock by validating if the
+	// disk is unformatted. If for some reason TryWriteLock fails
+	// and the error returned is lock.ErrAlreadyLocked i.e some other
+	// process is holding a lock we make sure that we try again
+	// in the loop by holding a read lock.
+	for {
+		var rlk *lock.RLockedFile
+		rlk, err = checkLockedValidFormatFS(fsPath)
```

Comment:
checkLockedValidFormatFS() can also return lock.ErrAlreadyLocked, shouldn't we also take care of that?